Create cloud vars

kchadha opened the pull request:
### Resolves

Towards having fully functional cloud variables in 3.0.

### Proposed Changes

Adds the ability to create cloud variables in a 3.0 project.
- adds a cloud variable option to the variable modal
- refactors cloud data connection logic into a separate HOC. 

#### Known Issues/TODO
- There is still more logic to come (not in this PR) about checking whether the user is a scratcher and about handling the cases where the user is viewing someone else's cloud project in editor mode.
- Adding a tooltip for why the cloud option becomes disabled (e.g. notifying the user about the cloud variable limit)
- Adding alerts/info bubbles about cloud variable usage when creating the first cloud variable in a project (like in 2.0)


### Reason for Changes

Cloud variables!

### Test Coverage

Manual testing:

On a local scratch environment (api, scratchr2, www, projects, assets), create a new account (doesn't yet have to be a scratcher) and ensure the following:
- [ ] navigate to /preview/editor to create a new project, create a cloud variable using the variable modal, see that the cloud variable gets created
- [ ] You should not be able to create more than 8 cloud variables (cloud var option becomes disabled in menu)
- [ ] Create code to modify the cloud variable, save, and share the project.
- [ ] Create another account and interact with the cloud project and see that the value of the cloud variable updates on both accounts (use separate browsers)
- [ ] In one browser, sign out and try to interact with the project (the cloud variable should neither send updates so that they show up on other accounts, nor should they receive updates made from another account).
- [ ] In signed out mode, in the editor, the cloud variable option should not show up in the variable modal
- [ ] Renaming a cloud variable should append a cloud + space if not present at the front of the rename, and should not append an extra one if it is present
- [ ] Cloud variable option should not show up in the variable modal in standalone gui

### Related PRs
- LLK/scratch-vm#1755
- LLK/scratch-blocks#1782

### Browser Coverage
Check the OS/browser combinations tested (At least 2)

Mac
 * [x] Chrome 
 * [x] Firefox 
 * [x] Safari
 
Windows
 * [ ] Chrome 
 * [ ] Firefox 
 * [ ] Edge
 
Chromebook
 * [ ] Chrome
 
iPad
* [ ] Safari

Android Tablet
* [ ] Chrome

paulkaplan commented on src/containers/blocks.jsx:
```diff
@@ -380,10 +380,12 @@ class Blocks extends React.Component {
             optVarType !== this.ScratchBlocks.BROADCAST_MESSAGE_VARIABLE_TYPE &&
             p.prompt.title !== this.ScratchBlocks.Msg.RENAME_VARIABLE_MODAL_TITLE &&
             p.prompt.title !== this.ScratchBlocks.Msg.RENAME_LIST_MODAL_TITLE;
+        // TODO CLOUD add flag about showing the cloud variable option, which will necessitate
```
hmmm? This is a confusing comment

paulkaplan commented on src/lib/cloud-provider.js:
```diff
@@ -65,6 +58,26 @@ class CloudProvider {
+            break;
+        }
+        }
+        this.vm.postIOData('cloud', varData);
```
Can you return the varData and put the postIOData in the caller of this function? Seems strange to have a method called "parse_" that actually sends data to the VM.

paulkaplan commented on src/lib/cloud-manager-hoc.jsx:
```diff
@@ -0,0 +1,98 @@
+            } = this.props;
+            return (
+                <WrappedComponent
+                    canUseCloud={this.canUseCloud}
```
It would be better to compute the value here instead of relying on a pre-computed value. This will ensure that the value of `canUseCloud` is always up-to-date given the props, whereas setting it in componentDidUpdate seems like it will always be one step behind.

kchadha replied:
@ray, yeah I think that makes sense. This is somewhat temporary because I needed to be able to control when/when not to show the cloud variable option in the variable modal in this set of PRs. I will probably move some of this logic into www when I add more `should-we-connect??` logic, but we should maybe talk offline about what's best to have where.

paulkaplan commented on src/lib/cloud-manager-hoc.jsx:
```diff
@@ -0,0 +1,98 @@
+                (this.props.projectId !== prevProps.projectId) ||
+                (!this.cloudProvider) ||
+                (this.cloudProvider && !this.cloudProvider.connection)) {
+                // TODO need to add provisions for viewing someone
```
Because this is in `componentDidUpdate`, it will not be called for the initial render. You probably want to extract lines 35-44 into a separate function that you can call here (during the update) and also in `componentDidMount`. I'm guessing this went unnoticed because the load process would never init the cloud server on mount because the project isn't loaded yet.

paulkaplan commented on src/lib/cloud-manager-hoc.jsx:
```diff
@@ -0,0 +1,98 @@
+                    this.props.vm,
+                    this.props.username,
+                    this.props.projectId);
+                this.props.vm.setCloudProvider(this.cloudProvider);
```
Can you add a componentDidMount that closes the websocket connection?

kchadha commented on src/lib/cloud-manager-hoc.jsx:
```diff
@@ -0,0 +1,98 @@
+
+    const mapDispatchToProps = () => ({});
+
+    // Allow incoming props to override redux-provided props. Used to mock in tests.
```
hahaha that was a copy paste remnant from the vm-manager-hoc =]

kchadha replied:
@rschamp, @paulkaplan, I think I will end up following up this PR with one that adds a test for the cloud manager hoc. I got stuck trying to figure out how to mock WebSocket (or maybe even CloudProvider) and might need some help on that.

kchadha replied:
JK, I ended up adding a test here after all.

kchadha commented on src/lib/cloud-manager-hoc.jsx:
```diff
@@ -0,0 +1,98 @@
+
+            if (!canUseCloud) return;
+
+            // TODO add scratcher check
```
I'm going to keep the comment in here for now as a reminder and will remove it when I use updated info from www in here.

kchadha commented:
@rschamp, @paulkaplan, I believe I addressed all the PR comments except for the addition of the test. I started work on this but got stuck trying to mock out CloudProvider/WebSocket, so I think I will leave the test for a follow up PR.

@paulkaplan, I modified the cloud-manager-hoc a bit further from when we worked on it together, notable changes are:
- use of componentWillUnmount instead of componentDidUnmount (when I was running the tests, they were complaining that componentDidUnmount is not a real lifestyle method, and it looks like elsewhere throughout the gui we are, in fact, using `componentWillUnmount`.
- the new `canUseCloud` function takes in props so that during componentDidUpdate, we can compare its value to what it would have been. This is important for the initial cloud connection so that it catches the change of loading state from isShowingWithId being false to being true. This is especially important when loading a project's page directly via `/preview/<project id>` for example.

kchadha commented:
@rschamp, @paulkaplan, I ended up figuring out how to mock a class via the awesome docs here:
https://jestjs.io/docs/en/es6-class-mocks

I added the test to this PR.
